File: webapp/src/lib/storage/local.ts
import fs from 'fs';
import StorageProvider from 'lib/storage/provider';
import path from 'path';
import util from 'util';
import { createLogger } from 'utils/logger';

const log = createLogger('webapp:storage:local');

let mkdir, unlink, writeFile;
if (typeof fs?.mkdir === 'function') {
	mkdir = util.promisify(fs.mkdir);
	unlink = util.promisify(fs.unlink);
	writeFile = util.promisify(fs.writeFile);
}

class LocalStorageProvider extends StorageProvider {
	static allowedDeleteErorCodes: string[] = ['ENOENT'];

	#basePath: string;

	constructor() {
		super();
		if (typeof fs?.mkdir !== 'function') {
			return;
		}
		this.#basePath = process.env.UPLOADS_BASE_PATH || './uploads';
		this.init();
	}

	async init() {
		try {
			await mkdir(this.#basePath, { recursive: true });
		} catch (e) {
			log.error(`Failed to create base directory: ${e.message}`);
			throw e;
		}
	}

	async uploadLocalFile(filename, uploadedFile, contentType, isPublic = false) {
		const filePath = path.join(this.#basePath, filename);
		try {
			await writeFile(filePath, uploadedFile.data);
			log.info(`File '${filename}' uploaded successfully.`);
		} catch (e) {
			log.error(`Failed to upload file: ${e.message}`);
			throw e;
		}
	}

	async uploadBuffer(
		filename: string,
		content: Buffer,
		contentType: string,
		isPublic = false
	): Promise<any> {
		const filePath = path.join(this.#basePath, filename);
		try {
			await writeFile(filePath, content);
			log.info(`Buffer uploaded to '${filename}' successfully.`);
		} catch (e) {
			log.error(`Failed to upload buffer: ${e.message}`);
			throw e;
		}
	}

	async cloneFile(sourceFilename: string, destinationFilename: string): Promise<any> {
		const sourceFilePath = path.join(this.#basePath, sourceFilename);
		const destinationFilePath = path.join(this.#basePath, destinationFilename);

		try {
			// Read the source file
			const fileContent = await util.promisify(fs.readFile)(sourceFilePath);

			// Write the content to the destination file
			await writeFile(destinationFilePath, fileContent);

			log.info(`File cloned successfully from '${sourceFilename}' to '${destinationFilename}'`);

			return destinationFilePath;
		} catch (err) {
			log.error(`Failed to clone file: ${err.message}`);
			throw err;
		}
	}

	async deleteFile(filename) {
		const filePath = path.join(this.#basePath, filename);
		try {
			await unlink(filePath);
			log.info(`File '${filename}' deleted successfully.`);
		} catch (e) {
			if (!LocalStorageProvider.allowedDeleteErorCodes.includes(e?.code)) {
				log.error(`Failed to delete file: ${e.message}`);
				throw e;
			}
		}
	}

	getBasePath() {
		return '/static';
	}
}

export default new LocalStorageProvider();

File: agent-backend/src/storage/local.py
import json
import os
import logging
from pathlib import Path
from init.env_variables import UPLOADS_BASE_PATH
from storage.provider import StorageProvider
from minio import Minio
from minio.error import S3Error

logging.basicConfig(level=logging.DEBUG)
log = logging.getLogger("storage.local")


class LocalStorageProvider(StorageProvider):
    allowed_delete_error_codes = ["ENOENT"]

    def __init__(self):
        self.base_path = UPLOADS_BASE_PATH if UPLOADS_BASE_PATH else "./uploads"
        self.init()
    def init(self):
        Path(self.base_path).mkdir(parents=True, exist_ok=True)

    def upload_file_buffer(self, buffer, filename, folder_path, is_public=False):
        log.debug('Uploading buffer content as file %s', filename)
        
        if not os.path.exists(os.path.join(self.base_path, folder_path)):
            os.makedirs(os.path.join(self.base_path, folder_path))
        
        file_path = os.path.join(self.base_path, folder_path, filename)
        
        try:
            with open(file_path, 'wb') as f:
                f.write(buffer.read())
            log.debug('Buffer content uploaded successfully.')
        except Exception as err:
            log.error('Buffer upload error:', err)
            raise err
    def delete_file(self, filename, file_folder, is_public=False):
        log.debug("Deleting file %s", filename)
        file_path = os.path.join(self.base_path, file_folder, filename)
        
        try:
            os.remove(file_path)
            log.debug("File deleted successfully.")
        except FileNotFoundError:
            log.warning("File not found: %s", file_path)
        except Exception as err:
            log.error("Error deleting file: %s", err)
            raise err

    def get_signed_url(self, filename, file_folder, is_public=False):
        return f"{os.getenv('WEBAPP_URL')}/tmp/{file_folder}/{filename}"



local_storage_provider = LocalStorageProvider()
